### DISCUSSION TITLE
Passing states and function between pages

### DISCUSSION THREAD

**Discussion by popeating** [vercel/next.js#20132] [Help]
I'm migrating a react app to Next js and i encountered the following stop;
it is a basic shopping cart and in my react js project i used react router
basically i have a page with a list of products with add to cart button, clicking the add to cart it also set an alert on the page to feedback the user about the added product.
it was something like:
```
const [addedToCart, setAddedToCart] = useState(false);
...
const addToCart = ({ product_id, qty }) => {
    const isPresent = cart.filter((pcart, productIndex) => {
      return pcart.id === product_id;
    });
    setIsLoading(true);

    const currentCatalog = [...products];
    const newCart = currentCatalog.filter((product, productIndex) => {
      return product.id === product_id;
    });
    newCart[0].in_cart = parseInt(qty);
    if (isPresent.length > 0) {
      isPresent[0].in_cart = isPresent[0].in_cart + parseInt(qty);

      setCart(....
    } else {
      setCart([...cart, ...newCart]);
    }

    setAddedToCart(true);
    setIsLoading(false);
  };
```
then every product have a link to the detail page:
```
<Link to={`product/${product.slug}`}>
                          {product.name}
                        </Link>
<Route path="/product/:slug">
              <Product
                cart={cart}
                addToCart={addToCart}
                addedToCart={addedToCart}
              />
            </Route>
```

this way i passed the product page the cart, the function to add the product to cart and the addedtocart state

now that im switching to Next, im puzzled on how to pass things around pages, i was able to replicate mostly of the structure, but for example, how can i pass the addToCart functions that i have i index to other pages? also sates (i know i can use context or Redux, but at the moment im keeping away from them)

Thank you

**Comment by Tomburgs** [comment]
You could create a custom [`App`](https://nextjs.org/docs/advanced-features/custom-app) and pass down the state / function to `Component` from there.

It seems, though, like you'll have to do a lot of prop-drilling to get the state down to the components which will actually use it.
Defining context there might be a better idea.

**Reply by popeating** [reply]
Thank you for your reply, its not much prop drilling to be honest, it's just a couple of function and a couple of state passing from App.js (home page) to Product.js (product page, made by a component in my React).
When you talk about custom App, you mean not using Next js page routing? (i already implemented navigation with it! especially i fixed all the dynamic page part. Also, how context can work on Next? usually context data is passed to child component, but using "page" structure my product page is not a child of my index page, should i wrap in context the _app.js?

**Reply by Tomburgs** [reply]
What do you mean by not using Next.js page routing?
Next.js automatically registers all the routes that are under `pages/` directory, `pages/_app.tsx` is simply a parent for every page.

Have a look at the [docs](https://nextjs.org/docs/advanced-features/custom-app), the `Component` prop it receives is the requested page.

As for context - yes, you would simply wrap your component within the context element in `App`.
You can have a look [here for an example](https://github.com/Tomburgs/pwa-boilerplate/blob/master/src/pages/_app.tsx#L36).

**Reply by popeating** [reply]
Your example is perfect
i just started today to convert my plain React to Next, so im still a bit confused about pages, now it make all more sense,

**Reply by popeating** [reply]
I thougth i was understanding it, but seems i did not
i made a (basic) context file
```
import React from 'react';

export const StoreContext = React.createContext(null);

const StoreProvider = ({ children }) => {
  const teamMembersNames = ['John', 'Mary', 'Jason', 'David'];

  const [sharing, setSharing] = React.useState([]);
  const [help, setHelp] = React.useState([]);
  const [pairing, setPairing] = React.useState(teamMembersNames);

  const store = {
    sharing: [sharing, setSharing],
    help: [help, setHelp],
    pairing: [pairing, setPairing],
  };

  return (
    <StoreContext.Provider value={store}>{children}</StoreContext.Provider>
  );
};
export default StoreProvider;
```
then in _app.js
```
import StoreProvider from '../lib/context';

function MyApp({ Component, pageProps }) {
  return (
    <>
      <StoreProvider>
        <Component {...pageProps} />
      </StoreProvider>
    </>
  );
}

export default MyApp;
```
and index.js
```
import StoreContext from '../lib/context';
export default function Home() {
const { isLoading } = useContext(StoreContext);
.....
```

but i got the following error
TypeError: Cannot read property 'isLoading' of undefined
(here -> const { isLoading } = useContext(StoreContext);)
what am i missing?

**Reply by Tomburgs** [reply]
You're passing the `StoreProvider` component (default export) to `useContext` instead of the `StoreContext` (named export).

**Reply by popeating** [reply]
Got it!

**Reply by Tomburgs** [reply]
> Got it!

If your issue is solved, feel free to set this discussion as answered :)
